src/encryptionkey.rs

```diff
@@ -68,51 +72,92 @@ impl EncryptionKey {
     where
         M: AsRef<[u8]>,
     {
-        let xx = BigNumber::from_slice(x);
-        if !mod_in(&xx, &self.n) {
+        let m = BigNumber::from_slice(x);
+        let r = r.unwrap_or_else(|| Nonce::random(&self.n));
+
+        if !mod_in(&m, &self.n) {
+            println!("m");
             return None;
         }
 
-        let r = r.unwrap_or_else(|| Nonce::random(&self.n));
-
         if !mod_in(&r, &self.n) {
+            println!("r");
             return None;
         }
 
-        // a = (N+1)^m mod N^2
-        let a = (&self.n + BigNumber::one()).modpow(&xx, &self.nn);
-        // b = r^N mod N^2
-        let b = &r.modpow(&self.n, &self.nn);
+        Some((self.encrypt_with_randomness(&m, &r), r))
+    }
+
+    /// Encrypt a given message with the encryption key and optionally use a random value
+    /// x must be less than N
+    #[allow(clippy::many_single_char_names)]
+    pub fn encrypt_unchecked<M>(&self, x: M, r: Option<Nonce>) -> (Ciphertext, Nonce)
+    where
+        M: AsRef<[u8]>,
+    {
+        let m = BigNumber::from_slice(x);
+        let r = r.unwrap_or_else(|| Nonce::random(&self.n));
+
+        (self.encrypt_with_randomness(&m, &r), r)
+    }
+
+    /// Encrypt a given message with the encryption key and a provided random value
+    /// x must be less than N
+    #[allow(clippy::many_single_char_names)]
+    pub fn encrypt_with_randomness(&self, m: &BigNumber, r: &Nonce) -> Ciphertext {
+        debug_assert!(mod_in(m, &self.n));
+
+        debug_assert!(mod_in(r, &self.n));
 
-        let c = a.modmul(&b, &self.nn);
-        Some((c, r))
+        // g^m mod N^2 = (N + 1)^m mod N^2 = m N + 1 mod N^2
+        // See Prop 11.26, Pg. 385 of Intro to Modern Cryptography
+        let g_m = m
+            .modmul(&self.n, &self.nn)
+            .modadd(&BigNumber::one(), &self.nn);
```

Comment:
Changed some math here

Reply:
Cool that saves a modpow.  Looks like it follows from binomial coefficient expansion with all the `N^k` terms vanishing for `k>1`.